python/ray/train/_internal/logging.py

```diff
@@ -0,0 +1,143 @@
+
+
+class SessionFileHandler(logging.Handler):
+    """A handler that writes to a log file in the Ray session directory.
+
+    The Ray session directory isn't available until Ray is initialized, so this handler
+    lazily creates the file handler when you emit a log record.
+
+    Args:
+        filename: The name of the log file. The file is created in the 'logs' directory
+            of the Ray session directory.
+    """
+
+    def __init__(self, filename: str):
+        super().__init__()
+        self._filename = filename
+        self._handler = None
+        self._formatter = None
+        self._path = None
+
+    def emit(self, record):
+        if self._handler is None:
+            self._try_create_handler()
+        if self._handler is not None:
+            self._handler.emit(record)
+
+    def setFormatter(self, fmt: logging.Formatter) -> None:
+        if self._handler is not None:
+            self._handler.setFormatter(fmt)
+        self._formatter = fmt
+
+    def _try_create_handler(self):
+        assert self._handler is None
+
+        log_directory = get_log_directory()
+        if log_directory is None:
+            return
+
+        os.makedirs(log_directory, exist_ok=True)
+
+        self._path = os.path.join(log_directory, self._filename)
+        self._handler = logging.FileHandler(self._path)
+        if self._formatter is not None:
+            self._handler.setFormatter(self._formatter)
+
+
+def configure_logging() -> None:
+    """Configure the Python logger named 'ray.train'.
+
+    This function loads the configration YAML specified by "RAY_TRAIN_LOGGING_CONFIG"
+    environment variable. If the variable isn't set, this function loads the default
+    "logging.yaml" file that is adjacent to this module.
+
+    If "RAY_TRAIN_LOG_ENCODING" is specified as "JSON" we will enable JSON reading mode
+    if using the default logging config.
+    """
+
+    def _load_logging_config(config_path: str):
+        with open(config_path) as file:
+            config = yaml.safe_load(file)
+        return config
+
+    if RAY_TRAIN_LOG_CONFIG_PATH is not None:
+        config = _load_logging_config(RAY_TRAIN_LOG_CONFIG_PATH)
+    else:
+        config = _load_logging_config(DEFAULT_RAY_TRAIN_LOG_CONFIG_PATH)
+        if RAY_TRAIN_LOG_ENCODING == RAY_TRAIN_JSON_LOG_ENCODING_FORMAT:
+            for logger in config["loggers"].values():
+                logger["handlers"].remove("file")
+                logger["handlers"].append("file_json")
+
+    logging.config.dictConfig(config)
+
+    # After configuring logger, warn if RAY_TRAIN_LOGGING_CONFIG_PATH is used with
+    # RAY_TRAIN_LOG_ENCODING, because they are not both supported together.
+    if RAY_TRAIN_LOG_CONFIG_PATH is not None and RAY_TRAIN_LOG_ENCODING is not None:
+        logger = logging.getLogger("ray.train")
+        logger.warning(
+            f"Using {RAY_TRAIN_LOG_ENCODING_ENV} is not supported with "
+            f"{RAY_TRAIN_LOG_CONFIG_PATH_ENV}."
```

Comment:
Add:

"If you are already specifying a custom config yaml file, please configure your desired encoding in the yaml file as well."